### ISSUE TITLE
Type inference incorrect with `extends keyof`

### ISSUE THREAD

**Issue by timsuchanek** [Microsoft/TypeScript#30319]
In this example
```ts
type User = {
  name: string
  age?: number
}

function a<T extends User>(param: T): 'age' extends keyof T ? number : string {
  if (param.age) {
    return param.age
  }

  return param.name
}

a({
  age: 1,
  name: '',
})
```
The TypeScript compiler (@next) is complaining:
```bash
src/type-inference.ts:8:5 - error TS2322: Type 'number' is not assignable to type '"age" extends keyof T ? number : string'.

8     return param.age
      ~~~~~~~~~~~~~~~~

src/type-inference.ts:11:3 - error TS2322: Type 'string' is not assignable to type '"age" extends keyof T ? number : string'.

11   return param.name
     ~~~~~~~~~~~~~~~~~
```
It should correctly infer though, that the `age` property exists in the `T` type, when asked for it in an `if`

**Comment by jack-williams** [comment]
The correspondence between keys at the value level and keys at the type level is not strong enough to make this sound. Consider this call:
```ts
const user: User = { name: "foo" };
const age: number = a(user);
```

**Comment by timsuchanek** [comment]
Thanks. That makes sense!
